website/programs.py

```diff
@@ -69,6 +71,11 @@ def store_user_program(self,
             "adventure_name": adventure_name,
         }
 
+        full_adventures = hedy_content.Adventures("en").get_adventures(g.keyword_lang)
+        teacher_adventures = self.db.get_teacher_adventures(current_user()["username"])
+        is_modified = self.statistics.is_program_modified(updates, full_adventures, teacher_adventures)
```

Comment:
Hi @Annelein! This is where the problem lies. We are sending this `updates` object and `is_program_modified` expects a `program`, specifically with the fields: `adventure_name`, `code`, and `level`. Now, I see that `adventure_name` might be null on this function, which can cause issues in the function, so I'll say that is `adventure_name` is null we just return true and save the program.

Reply:
> Hi @Annelein! This is where the problem lies. We are sending this `updates` object and `is_program_modified` expects a `program`, specifically with the fields: `adventure_name`, `code`, and `level`. Now, I see that `adventure_name` might be null on this function, which can cause issues in the function, so I'll say that is `adventure_name` is null we just return true and save the program.

@jpelay but this already returns True if `adventure_name` is null, because then `if program['adventure_name'] == adventure["id"]:` does not happen right?

https://github.com/hedyorg/hedy/blob/dbc117abc336199be2bb87af45f6d8d0e36554a6/website/statistics.py#L267C1-L277C43

Reply:
And level and code will never be null, so I still don't really get why copied code is being saved as a program?

Reply:
what does this line do?         
```
adventure_info = full_adventures.get(program['adventure_name'], {})\
            .get('levels', {})\
            .get(program['level'], {})
```

because it becomes None / empty, while adventure_name and level are set in program:
`{'session': '8179ba4ca4e242e29959017837f4313b', 'date': 1712750025338, 'lang': 'en', 'version': 'DEV', 'level': 1, 'code': 'forward 100\nturn left\n', 'name': 'Turtle 1', 'username': 'teacher1', 'error': False, 'adventure_name': 'Turtle'} `

Reply:
This line get's the adventure information. It's a dictionary where each adventure is identified by its key in the yalm. I see in the example you posted that `adventure_name` has the value `Turtle` which is not the key of the adventure, but the name in English, the real key is: `turtle` that's why is returning empty.